src/main/resources/assets/carpet/scripts/shapes.scl

```diff
@@ -97,134 +132,178 @@ draw_prism(args)->(
 
     positions = {};
 
-    positions += fill_flat_circle_points(pos, orientation, offset, hollow, a, b);
+    if(r == 0,
+        offset_vector = if(orientation=='x', [1,0,0], orientation=='y', [0,1,0],[0,0,1]);
+        positions += pos + offset_vector*offset;
+        return(positions)
+    );
+
+    a=0;
+    b=r;
+    d = 1.25-r;
 
-    while(a<=b, 10^10,
-        if(d<=0,
-            d += (4*a)+6,
-            d += (4*a)-(4*b)+10;
-            b=b-1
+    while(a<=b, r, //technically the limit is r/sqrt(2) but whatevs
+        positions = positions + _fill_flat_circle_points(pos, orientation, offset, hollow, a, b);
+        if(d > 0,
+	        d+=2*a-2*b+5;
+	        b=b-1;
+	        if(hollow && thick, //to make circles slightly thicker
+	            positions = positions + _fill_flat_circle_points(pos, orientation, offset, hollow, a, b)
+	        ),
+	        d+=2*a+3
         );
         a+=1;
-        positions += fill_flat_circle_points(pos, orientation, offset, hollow, a, b)
     );
     positions
 );
 
-fill_flat_circle_points(pos, orientation, offset, hollow, a, b) ->(
+//Method used to fill points in 8 octants.
+//Used by Bresenham midpoint algorithm to use least code,
+//exploiting the symmetricity of the circle.
+//Algorithmic efficiency:
+//  O(1) if hollow cos it's just 8 points
+//  O(n) if filled cos it draws the lines between the points
+_fill_flat_circle_points(pos, orientation, offset, hollow, a, b) ->(
     [x, y, z] = pos;
     positions = {};
     if(orientation == 'x',
         if(hollow,
-            positions += [x + offset, a+y, b+z];
-            positions += [x + offset, a+y,-b+z];
-            positions += [x + offset,-a+y,-b+z];
-            positions += [x + offset,-a+y, b+z];
-            positions += [x + offset, b+y, a+z];
-            positions += [x + offset, b+y,-a+z];
-            positions += [x + offset,-b+y,-a+z];
-            positions += [x + offset,-b+y, a+z],
-            c_for(c = -a, c<=a, c+=1,
-                positions += [x + offset,c+y, b + z];
-                positions += [x + offset,c+y,-b + z]
+            positions += [x + offset,y+a,z+b];
+            positions += [x + offset,y+a,z-b];
+            positions += [x + offset,y-a,z+b];
+            positions += [x + offset,y-a,z-b];
+            positions += [x + offset,y+b,z+a];
+            positions += [x + offset,y+b,z-a];
+            positions += [x + offset,y-b,z+a];
+            positions += [x + offset,y-b,z-a],
+            for(range(-b, b+1),
+                positions += [x + offset,y+a,_ + z];
+                positions += [x + offset,y-a,_ + z]
             );
-            c_for(c = -b, c<=b, c+=1,
-                positions += [x + offset, a+y,c + z];
-                positions += [x + offset,-a+y,c + z]
+            for(range(-a, a+1),
+                positions += [x + offset,y+b,_ + z];
+                positions += [x + offset,y-b,_ + z]
             )
         ),
         orientation=='y',
         if(hollow,
```

Comment:
IMO this is very difficult to realize that it's changing branch in an `if`, been confused for a while in why it was repeated.

Reply:
Wdym? If ur asking about the orientations thing (where I have a different branch for x orientation, y orientation and z orientation), that was an old decision by gnembon, and I fully agree with it.